server/src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java

```diff
@@ -139,7 +140,10 @@ private StoreFilesMetaData listStoreMetaData(ShardId shardId) throws IOException
                 return new StoreFilesMetaData(shardId, Store.MetadataSnapshot.EMPTY);
             }
             final IndexSettings indexSettings = indexService != null ? indexService.getIndexSettings() : new IndexSettings(metaData, settings);
-            final ShardPath shardPath = ShardPath.loadShardPath(logger, nodeEnv, shardId, indexSettings);
+            final ShardPath shardPath;
+            try (ShardLock ignored = nodeEnv.shardLock(shardId, TimeUnit.SECONDS.toMillis(5))) {
```

Comment:
did you double check what the effect is of failing to get the lock?

Reply:
In `TransportNodesListGatewayStartedShards` and in `Store.readMetadataSnapshot`, which we call below, we catch the `ShardLockObtainFailedException` and treat it either as an empty store (in case of  `TransportNodesListShardStoreMetaData`) or as an ok target for primary allocation (see `TransportNodesListGatewayStartedShards` and `PrimaryShardAllocator.buildNodeShardsResult`), but we've made sure not to end up in a situation where the master goes into a potentially long retry loop (which causes a reroute storm on the master). I don't want to open this box of Pandora here, so my suggestion is to add
```
} catch (ShardLockObtainFailedException ex) {
    logger.info(() -> new ParameterizedMessage("{}: failed to obtain shard lock", shardId), ex);
    return new StoreFilesMetaData(shardId, Store.MetadataSnapshot.EMPTY);
}
```
here so as not to mess with existing behavior.